ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -72,3 +158,203 @@ func (r *RayServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if errStatus := r.Status().Update(context.Background(), rayServiceInstance); errStatus != nil {
+		return fmt.Errorf("combined error: %v %v", err, errStatus)
+	}
+	return err
+}
+
+func (r *RayServiceReconciler) getOrCreateRayClusterInstance(rayServiceInstance *rayv1alpha1.RayService) (*rayv1alpha1.RayCluster, error) {
+	// Update ray cluster
+	rayClusterNamespacedName := types.NamespacedName{
+		Namespace: rayServiceInstance.Namespace,
+		Name:      rayServiceInstance.Name + rayClusterSuffix,
+	}
+
+	rayClusterInstance := &rayv1alpha1.RayCluster{}
+	err := r.Get(context.TODO(), rayClusterNamespacedName, rayClusterInstance)
+
+	if err == nil {
+		rayClusterInstance.Spec = rayServiceInstance.Spec.RayClusterSpec
+
+		r.Log.Info("Update ray cluster spec")
+		if err := r.Update(context.TODO(), rayClusterInstance); err != nil {
+			r.Log.Error(err, "Fail to update ray cluster instance!")
+			// Error updating the RayCluster object.
+			return nil, client.IgnoreNotFound(err)
+		}
+	} else if errors.IsNotFound(err) {
+		r.Log.Info("Not found rayCluster, creating rayCluster!")
+		rayClusterInstance, err = r.constructRayClusterForRayService(rayServiceInstance)
+		if err != nil {
+			r.Log.Error(err, "unable to construct rayCluster from spec")
+			// Error construct the RayCluster object - requeue the request.
+			return nil, err
+		}
+		if err := r.Create(context.TODO(), rayClusterInstance); err != nil {
+			r.Log.Error(err, "unable to create rayCluster for rayService", "rayCluster", rayClusterInstance)
+			// Error creating the RayCluster object - requeue the request.
+			return nil, err
+		}
+		r.Log.V(1).Info("created rayCluster for rayService run", "rayCluster", rayClusterInstance)
+	} else {
+		r.Log.Error(err, "Get request rayCluster instance error!")
+		// Error reading the RayCluster object - requeue the request.
+		return nil, client.IgnoreNotFound(err)
+	}
+
+	return rayClusterInstance, nil
+}
+
+func (r *RayServiceReconciler) constructRayClusterForRayService(rayService *rayv1alpha1.RayService) (*rayv1alpha1.RayCluster, error) {
+	name := fmt.Sprintf("%s%s", rayService.Name, rayClusterSuffix)
+
+	rayCluster := &rayv1alpha1.RayCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      rayService.Labels,
+			Annotations: rayService.Annotations,
+			Name:        name,
+			Namespace:   rayService.Namespace,
+		},
+		Spec: *rayService.Spec.RayClusterSpec.DeepCopy(),
+	}
+
+	// Set the ownership in order to do the garbage collection by k8s.
+	if err := ctrl.SetControllerReference(rayService, rayCluster, r.Scheme); err != nil {
+		return nil, err
+	}
+
+	return rayCluster, nil
+}
+
+func (r *RayServiceReconciler) fetchDashboardURL(rayCluster *rayv1alpha1.RayCluster) (string, error) {
+	headServices := corev1.ServiceList{}
+	filterLabels := client.MatchingLabels{common.RayClusterLabelKey: rayCluster.Name}
+	if err := r.List(context.TODO(), &headServices, client.InNamespace(rayCluster.Namespace), filterLabels); err != nil {
```

Comment:
Would it be possible to acquire the head service by a "get" rather than a "list"? We should be able to know the name of the head service ahead of time.

My fear of `List` operations might be irrational: It stems from experience making list calls without resource versions with the Python K8s client, which doesn't have caches to protect List performance.
https://kubernetes.io/docs/reference/using-api/api-concepts/#the-resourceversion-parameter

cc @Jeffwan @wilsonwang371 @akanso for thoughts on list vs. get.

Reply:
I think we should be able to use `get` directly. While there is a similar logic in RayCluster controller too. Let's align which one we want to use.

Reply:
Get is preferred when convenient! (It's less expensive than list.)